<!DOCTYPE html>
<html>
<head>
<meta charset="utf-8" />   <!--It is necessary to use the UTF-8 encoding with plotly graphics to get e.g. negative signs to render correctly -->
<meta name="viewport" content="width=device-width, initial-scale=1.0" />
<style>
    body {background-color: lightslategray; font-family: sans-serif; font-size: 10pt; }
    .bioraum {  display: block; font-size: 10px; color: grey ; background-color: lightgrey; padding: 5px; }
    a {color: slategray; text-decoration: none;}
    a:hover {color: darkgrey ; text-decoration: none;}
    h1 strong { display: block; font-size: 75%; opacity: 0.25; }
    h2 {background: lightslategray; color: white; padding-left: 5px ;}
    .ground {background: white ; display: block; padding: 5px; height: fit-content}
    .impressum {background: lightgrey; display: block; padding: 5px; vertical-align: bottom; font-size: 10px;}
    td, td, th { padding: 5px;}
    tr td[scope="row"] {background-color: lightgray;}
    tr td[scope="sample-id"] {background-color: lightgray; font-weight: bold;}
    tr td[scope="warn"] {background-color: rgb(240, 184, 80);}
    tr td[scope="pass"] {background-color: rgb(65, 156, 65); color: whitesmoke;} 
    tr td[scope="fail"] {background-color: rgb(236, 74, 74);}
    tr td[scope="missing"] {background-color: lightgreen;}
    tr th[scope="col"] {background-color: #505050; color: #fff;}
    tr th[scope="subcol"] {background-color: #888686; color: #fff;}
    .nav td { border-top: 1px solid; border-bottom: 1px solid ;}
    .versions { font-size: 9px; }
    .versions td { padding: 0px; }
    /* Tooltip container */
    .tooltip { position: relative; display: inline-block; border-bottom: 1px dotted black; }
    .tooltip .tooltiptext { visibility: hidden; width: 250px; background-color: darkgrey; color: black; text-align: center; padding: 5px 0; border-radius: 6px;  position: absolute;  z-index: 1; }
    .tooltip:hover .tooltiptext { visibility: visible; }
</style>
</head>

<body>

<div class="bioraum">
    <a href="https://github.com/bio-raum">github.com/bio-raum</a> - bioinformatic resources for surveillance in food safety and public health. 
</div>

<div class="ground">

<h1>
    <div id="summary"></div>GABI
    <strong>Genomic Analysis of Bacterial Isolates</strong>
</h1>

<div id="navigation"></div>
<table  class="nav">
    <tr>
        <td><a href="#summary">Summary</a></td>
        <td><a href="#assembly">Assembly</a></td>
        <td><a href="#kraken">Kraken2</a></td>
        <td><a href="#software">Software</a></td>
    </tr>
</table>
<!--
The summary table with all key metrics per sample
-->
<h2>Summary</h2>
<table>
    <tr>
        <th scope="col">Sample</th>
        <th scope="col"><div class="tooltip">Status<span class="tooltiptext">The overall analysis status: pass: ok to use, warn: potential issues found, fail: most probably not usable</span></div></th>
        <th scope="col"><div class="tooltip">Best-guess taxon<span class="tooltiptext">The highest scoring taxon in the Kraken2 analysis - green: robust call, orange: weak call, red: very weak call</span></div></th>
        <th scope="col"><div class="tooltip">Assembly (Mb)<span class="tooltiptext">The size of the chromosomal assembly (i.e. without plasmids). If reference sizes are configured, colors indicate concordance (green: within expected range, orange: <= 20% deviation, red: more than 20% deviation)</span></div></th>
        <th scope="col"><div class="tooltip">Chromosomal contigs<span class="tooltiptext">The number of chromosomal contigs (i.e. without plasmids) - smaller is better</span></div></th>
        <th scope="col"><div class="tooltip">N50 (Kb)<span class="tooltiptext">The assembly N50 (size of contigs that account for 50% of the total assembly size - bigger is better). If a minimum N50 is configured, colors indicate concordance (green = above minimum, orange: up to 20% below minimum, red: more than 20% below minimum). </span></div></th>
        <th scope="col"><div class="tooltip">Completeness (%)<span class="tooltiptext">Estimated completeness of the assembly, measured against a best-guess reference genome</span></div></th>
        <th colspan=4 scope="col"><div class="tooltip">Mean coverage<span class="tooltiptext">Mean coverage of reads mapped back to the assembly - bigger is better</span></div></th>
        <th scope="col"><div class="tooltip">Read contamination (%)<span class="tooltiptext">Highest observed predicted contamination across all read sets (Illumina and Pacbio)</span></div></th>
        <th scope="col"><div class="tooltip">Taxa above 10%<span class="tooltiptext">How many taxa were found by Kraken with >= 10% abundance - might indicate contamination issue</span></div></th>
    </tr>
    <tr>
        <th scope="col"></th>
        <th scope="col"></th>
        <th scope="col"></th>
        <th scope="col"></th>
        <th scope="col"></th>
        <th scope="col"></th>
        <th scope="col"></th>
        <th scope="subcol">Total</th>
        <th scope="subcol">Illumina</th>
        <th scope="subcol">ONT</th>
        <th scope="subcol">HiFi</th>
        <th scope="col"></th>
        <th scope="col"></th>
    </tr>
<!--
Our data structure is a list of dictionaries; where each dictionary holds
all the information for one sample    
-->
{% for row in summary %}
    <tr>
        <td scope="sample-id">{{row.sample}}</td>
        <td scope={{row.status}}>{{row.status}}</td>
        <td scope={{row.taxon_status}}>{{row.taxon}}</td>
        <td scope={{row.assembly_status}}>{{row.assembly}}</td>
        <td scope={{row.contigs_status}}>{{row.contigs}}</td>
        <td scope={{row.n50_status}}>{{row.n50}}</td>
        <td scope={{row.fraction_status}}>{{row.fraction}}</td>
        <td scope={{row.coverage_status}}>{{row.coverage}}</td>
        <td scope={{row.coverage_illumina_status}}>{{row.coverage_illumina}}</td>
        <td scope={{row.coverage_nanopore_status}}>{{row.coverage_nanopore}}</td>
        <td scope={{row.coverage_pacbio_status}}>{{row.coverage_pacbio}}</td>
        <td scope={{row.confindr_status}}>{{row.contamination}}</td>
        <td scope={{row.taxon_count_status}}>{{row.taxon_count}}</td>
    </tr>
{% endfor %}
</table>

<a href="#navigation">top</a>
<!--
The assembly metrics computed by QUAST
-->
<div id="assembly"></div>
<h2>Assembly metrics</h2>

<table>
    <tr>
        <th scope="col">Sample</th>
        <th scope="col">Assembly size</th>
        <th scope="col">Fraction of reference</th>
        <th scope="col">Ns per 100kb</th>
        <th scope="col">Largest contig</th>
        <th scope="col">Misassembled contigs</th>
    </tr>

{% for row in summary %}
    <tr>
        <td scope="sample-id">{{row.sample}}</td>
        <td scope="col">{{row.assembly}}</td>
        <td scope="col">{{row.fraction}}</td>
        <td scope="col">{{row.quast.N}}</td>
        <td scope="col">{{row.quast.largest_contig}}</td>
        <td scope="col">{{row.quast.misassembled}}</td>
    </tr>
{% endfor %}

</table>

<a href="#navigation">top</a>

<div id="kraken"></div>
<h2>Kraken2 - taxonomic composition</h2>

{{Kraken}}

<a href="#navigation">top</a>

<div id="software"></div>
<h2>Software versions</h2>

<table class="versions">
    
    {% for pmod,packs in packages.items() %}
        <tr>
            <td scope="col" colspan="2"><b>{{pmod}}</b></td>
        </tr>
        {% for pack in packs %}
            <tr>
                <td></td>
                <td scope="col">{{pack}}</td>
            </tr>
        {% endfor %}
    {% endfor %}
        </table>
<p></p>
<a href="#navigation">top</a>
</div>

<div class="impressum">
    Report generated by bio-raum/gabi. Please check out our <a href="https://github.com/bio-raum/gabi/blob/main/docs/usage.md">documentation</a>. 
</div>
</body>
</html>