<!DOCTYPE html>
<html>
<head>
<meta charset="utf-8" />   <!--It is necessary to use the UTF-8 encoding with plotly graphics to get e.g. negative signs to render correctly -->
<meta name="viewport" content="width=device-width, initial-scale=1.0" />
<style>
    body {background-color: lightslategray; font-family: sans-serif; font-size: 10pt; }
    .bioraum {  display: block; font-size: 10px; color: grey ; background-color: lightgrey; padding: 5px; }
    a {color: slategray; text-decoration: none;}
    a:hover {color: darkgrey ; text-decoration: none;}
    h1 strong { display: block; font-size: 75%; opacity: 0.25; }
    h2 {background: lightslategray; color: white; padding-left: 5px ;}
    .nav td { border-top: 1px solid; border-bottom: 1px solid ; border-left: 5px solid white;}
    .ground {background: white ; display: block; padding: 5px; height: fit-content}
    .impressum {background: lightgrey; display: block; padding: 5px; vertical-align: bottom; font-size: 10px;}
    .general { border-style: none; border-spacing: 0px ;}
    .table-caption { font-weight: bold; display: block;}
    table { border-collapse: collapse;}
    tr.row td { border-bottom: 1px solid grey; padding-top: 5px; padding-bottom: 5px;}
    tr.row-sample { cursor:pointer; }
    tr.row-sample td { border-bottom: 1px solid grey; padding-top: 5px; padding-bottom: 5px; }
    tr.row-hidden { display: none; text-align: left; }
    td { padding-left: 5px; padding-right: 5px;}
    th { border-left: 1px solid white; padding-right: 5px;}
    tr td[scope="sample-id"] {background-color: rgb(233, 233, 233); font-weight: bold;}
    tr td[scope="warn"] {background-color: rgb(255, 227, 195); border-left: 4px solid rgb(240, 184, 80);}
    tr td[scope="pass"] {background-color: rgb(204, 255, 206); border-left: 4px solid rgb(65, 156, 65);} 
    tr td[scope="fail"] {background-color: rgb(248, 187, 187); border-left: 4px solid rgb(236, 74, 74);}
    tr td[scope="missing"] {background-color: rgb(233, 233, 233); border-left: 4px solid lightgray;}
    tr th[scope="col"] {background-color: #505050; color: #fff; padding-left: 5px;}
    tr th[scope="subcol"] {background-color: #888686; color: #fff; padding-left: 5px;}
    tr th[scope="subcol-inactive"] {background-color: #6e6e6e; color: #fff;}
    tr td[scope="even"] {background-color: #eeeeee; color: #505050; border-left: 2px solid rgb(255, 255, 255); padding-left: 5px;}
    tr td[scope="uneven"] {background-color: #d9d8d8; color: #505050; padding-left: 5px;}
    tr td[scope="col_amr"] {background-color: #e2ffd7; color: #505050; padding-left: 5px;}
    tr td[scope="col_mlst"] {background-color: hsl(213, 100%, 91%); color: #505050; padding-left: 5px;}
    tr td[scope="col_sero"] {background-color: #fae1ed; color: #505050; padding-left: 5px;}
    .versions { font-size: 9px; }
    .versions td { padding: 0px; }
    .infotext { color:#6e6e6e; font-size: 10px ;}
    .infotext a { text-decoration: underline;}
    /* Tooltip container */
    .tooltip { position: relative; display: inline-block; border-bottom: 1px dotted black; }
    .tooltip .tooltiptext { visibility: hidden; width: 250px; background-color: rgb(189, 186, 186); color: black; text-align: center; padding: 5px 0; border-radius: 6px;  position: absolute;  z-index: 1; }
    .tooltip:hover .tooltiptext { visibility: visible; }

</style>
</head>

<body>

<div class="bioraum">
    <a href="https://github.com/bio-raum">github.com/bio-raum</a> - bioinformatic resources for surveillance in food safety and public health. 
</div>

<div class="ground">

<h1>
    <div id="summary"></div>GABI
    <strong>Genomic Analysis of Bacterial Isolates</strong>
</h1>

<div id="navigation"></div>
<table  class="nav">
    <tr>
        <td><a href="#summary">Summary</a></td>
        <td><a href="#assembly">Assembly</a></td>
        {% if Insertsizes %}
            <td><a href="#insert-sizes">Insert sizes</a></td>
        {% endif %}
        <td><a href="#busco">Busco</a></td>
        {% if Bracken_ILLUMINA %}
            <td><a href="#bracken">Bracken (Illumina)</a></td>
        {% endif %}
        {% if Bracken_NANOPORE %}
            <td><a href="#bracken_ont">Bracken (Nanopore)</a></td>
        {% endif %}
        {% if Bracken_PACBIO %}
            <td><a href="#bracken_pacbio">Bracken (Pacbio)</a></td>
        {% endif %}
        {% if serotypes %}
            <td><a href="#character">Characterization</a></td>
        {% endif %}
        <td><a href="#software">Software</a></td>
    </tr>
</table>

<!--
Some basic information about the analysis run
-->
<h2>Run Infos</h2>
<table class="general">
    <tr>
        <td>User</td><td>{{user}}</td>
    </tr>
    <tr>
        <td>Date</td><td>{{date}}</td>
    </tr>
    <tr>
        <td>Pipeline version</td><td>{{version}}</td>
    </tr>
    <tr>
        <td>Command line call</td><td>{{call}}</td>
    </tr>
    <tr>
        <td>Work directory</td><td>{{wd}}</td>
    </tr>
</table>

<!--
The summary table with all key metrics per sample
-->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js"></script>

<h2>Summary</h2>
<table class="general">
    <tr>
        <th scope="col">Sample</th>
        <th scope="col"><div class="tooltip">Status<span class="tooltiptext">The overall analysis status: pass: ok to use, warn: potential issues found, fail: most probably not usable</span></div></th>
        <th scope="col"><div class="tooltip">Best-guess taxon<span class="tooltiptext">The highest scoring taxon using kmer matching (S/MASH)</span></div></th>
        <th colspan=2 scope="col"><div class="tooltip">Reference genome<span class="tooltiptext">The highest matching hit in RefSeq to this assembly</span></div></th>
        <th colspan=6 scope="col"><div class="tooltip">Assembly<span class="tooltiptext">Information about this assembly</span></div></th>
        <th colspan=4 scope="col"><div class="tooltip">Mean coverage (X)<span class="tooltiptext">Mean coverage of reads mapped back to the assembly - bigger is better</span></div></th>
        <th colspan=4 scope="col"><div class="tooltip">Coverage 40X (%)<span class="tooltiptext">Percentage of assembly covered at 40X or more</span></div></th>
        <th colspan=3 scope="col"><div class="tooltip">Read quality<span class="tooltiptext">Quality metrics of reads after trimming</span></div></th>
        <th colspan=2 scope="col"><div class="tooltip">ConfindR<span class="tooltiptext">Indicators of contamination from sequence variation</span></div></th>
        <th colspan=2 scope="col"><div class="tooltip">Taxa<span class="tooltiptext">Number of detected taxa with an abundance of > 5% in the reads</span></div></th>
        <th colspan=2 scope="col"><div class="tooltip">Assembly Composition<span class="tooltiptext">Contamination check of the full assembly by checking taxonomic signature of all contigs</span></div></th>
    </tr>
    <tr class="row">
        <th scope="subcol-inactive"></th>
        <th scope="subcol-inactive"></th>
        <th scope="subcol-inactive"></th>
        <th scope="subcol">Assembly</th>
        <th scope="subcol"><div class="tooltip">% Cov<span class="tooltiptext">How much of the reference is covered by this assembly in percent.</span></div></th>
        <th scope="subcol">Size (Mb)</th>
        <th scope="subcol"><div class="tooltip">#Contigs<span class="tooltiptext">The number of chromosomal contigs, i.e. without plasmids.</span></div></th>
        <th scope="subcol"><div class="tooltip">N50 (Kb)<span class="tooltiptext">The size of contigs (>=)in which 50% of the assembly are represented.</span><div></th>
        <th scope="subcol"><div class="tooltip">Gene space (%)<span class="tooltiptext">The fraction of broadly conserved genes covered in this assembly (BUSCO), complete (C) | duplicated (D).</span></div></th>
        <th scope="subcol"><div class="tooltip">GC (%)<span class="tooltiptext">GC content of the assembly. Deviations from the species default are highlighted in orange (mild) and red (strong, something likely wrong)</span></div></th>
        <th scope="subcol"><div class="tooltip">Dup<span class="tooltiptext">Duplication ratio as measured against the reference. Deviations from the species default are highlighted in orange (mild) and red (strong, something likely wrong)</span></div></th>
        <th scope="subcol">Total</th>
        <th scope="subcol">ILM</th>
        <th scope="subcol">ONT</th>
        <th scope="subcol">HiFi</th>
        <th scope="subcol"><div class="tooltip">Total<span class="tooltiptext">Across all sequencing technologies</span></div></th>
        <th scope="subcol"><div class="tooltip">ILM<span class="tooltiptext">Illumina reads 40X</span></div></th>
        <th scope="subcol"><div class="tooltip">ONT<span class="tooltiptext">ONT reads 40X</span></div></th>
        <th scope="subcol"><div class="tooltip">HiFi<span class="tooltiptext">Pacbio reads 40X</span></div></th>
        <th scope="subcol"><div class="tooltip">ILM Q30 (%)<span class="tooltiptext">Fraction of Illumina reads above Q30.</span></div></th>
        <th scope="subcol"><div class="tooltip">ONT Q15 (#)<span class="tooltiptext">Number of ONT reads above Q15.</span></div></th>
        <th scope="subcol"><div class="tooltip">ONT N50 (bp)<span class="tooltiptext">N50 of ONT reads</span></div></th>
        <th scope="subcol"><div class="tooltip">ILM<span class="tooltiptext">Contaminating SNPs in Illumina reads.</span></div></th>
        <th scope="subcol"><div class="tooltip">ONT<span class="tooltiptext">Contaminating SNPs in ONT reads; not 100% reliable.</span></div></th>
        <th scope="subcol">ILM</th>
        <th scope="subcol">ONT</th>
        <th scope="subcol"><div class="tooltip">Contam %<span class="tooltiptext">CheckM2 prediction of contamination</span></div></th>
        <th scope="subcol"><div class="tooltip">Majority Genus %<span class="tooltiptext">Fraction of assembly assigned to the majority genus</span></div></th>
    </tr>
<!--
Our data structure is a list of dictionaries; where each dictionary holds
all the information for one sample    
-->
<script>
        $(document).ready(function(){
        $('tr.row-sample').click(function(){
            $(this).nextUntil('tr.row-sample').slideToggle(100, function(){
            });
        });
    });
</script>

{% for row in summary %}
    <tr class="row-sample">
        <td scope="sample-id">{{row.sample}}</td>
        <td scope={{row.status}}>{{row.status}}</td>
        <td scope="missing">{{row.taxon}}</td>
        {% if row.reference.assembly %}
            <td scope="missing"><a href="https://www.ncbi.nlm.nih.gov/datasets/genome/{{row.reference.assembly}}" target="_blank" title="{{row.reference.definition}}" alt="Genome in GenBank">{{row.reference.assembly}}</a></td>
        {% else %}
            <td scope="missing"><a href="https://www.ncbi.nlm.nih.gov/nuccore/{{row.reference.locus}}" target="_blank" title="{{row.reference.definition}}" alt="Genome in GenBank">{{row.reference.locus}}</a></td>
        {% endif %}
        <td scope={{row.fraction_status}}>{{row.fraction}}</td>
        <td scope={{row.assembly_status}}>{{row.assembly}}</td>
        <td scope={{row.contigs_status}}>{{row.contigs}}</td>
        <td scope={{row.n50_status}}>{{row.n50}}</td>
        <td scope={{row.busco_status}}>C: {{row.busco.completeness}} | D: {{row.busco.duplicated}}</td>
        <td scope={{row.quast.gc_status}}>{{row.quast.gc}}</td>
        <td scope={{row.quast.duplication_status}}>{{row.quast.duplication_ratio}}</td>
        <td scope={{row.coverage_status}}>{{row.coverage}}</td>
        <td scope={{row.coverage_illumina_status}}>{{row.coverage_illumina}}</td>
        <td scope={{row.coverage_nanopore_status}}>{{row.coverage_nanopore}}</td>
        <td scope={{row.coverage_pacbio_status}}>{{row.coverage_pacbio}}</td>
        <td scope={{row.coverage_40_status}}>{{row.coverage_40}}</td>
        <td scope={{row.coverage_40_illumina_status}}>{{row.coverage_40_illumina}}</td>
        <td scope={{row.coverage_40_nanopore_status}}>{{row.coverage_40_nanopore}}</td>
        <td scope={{row.coverage_40_pacbio_status}}>{{row.coverage_40_pacbio}}</td>
        <td scope={{row.quality_illumina_status}}>{{row.quality_illumina}}</td>
        <td scope="missing">{{row.quality_nanopore}}</td>
        <td scope="missing">{{row.nanopore_n50}}</td>
        <td scope={{row.confindr_illumina_status}}>{{row.contamination_illumina}}</td>
        <td scope={{row.confindr_nanopore_status}}>{{row.contamination_nanopore}}</td>
        <td scope={{row.taxon_count_illumina_status}}>{{row.taxon_count_illumina}}</td>
        <td scope={{row.taxon_count_nanopore_status}}>{{row.taxon_count_nanopore}}</td>
        <td scope={{row.checkm_status}}>{{row.checkm}}</td>
        <td scope={{row.taxonkit_genus_status}}>{{row.taxonkit_genus}}</td>
    </tr>
    <tr class="row-hidden">
        <th colspan="27">{{row.messages}}</th>
    </tr>
{% endfor %}
</table>
<p></p>
<a href="#navigation">top</a>

<!--
The assembly metrics computed by QUAST
-->
<div id="assembly"></div>
<h2>Assembly metrics</h2>
<div class="infotext">
   Descriptive metrics of individual assemblies determined by <a href="https://quast.sourceforge.net/" target="_new">Quast</a>. 
   <p></p>
</div>

<table>
    <tr>
        <th scope="col">Sample</th>
        <th scope="col">Assembly size (Mb)</th>
        <th scope="col">Fraction of reference</th>
        <th scope="col">Ns per 100kb</th>
        <th scope="col">Largest contig (Kb)</th>
        <th scope="col">Misassembled contigs</th>
        <th scope="col">Contigs > 1kb</th>
        <th scope="col">Contigs > 5kb</th>
        <th scope="col">Size (Mb) in contigs > 1kb</th>
        <th scope="col">Size (Mb) in contigs > 5kb</th>
    </tr>

{% for row in summary %}
    <tr>
        <td scope="sample-id">{{row.sample}}</td>
        <td scope="col">{{row.assembly}}</td>
        <td scope="col">{{row.fraction}}</td>
        <td scope="col">{{row.quast.N}}</td>
        <td scope="col">{{row.quast.largest_contig}}</td>
        <td scope="col">{{row.quast.misassembled}}</td>
        <td scope="col">{{row.quast.contigs_1k}}</td>
        <td scope="col">{{row.quast.contigs_5k}}</td>
        <td scope="col">{{row.quast.size_1k}}</td>
        <td scope="col">{{row.quast.size_5k}}</td>
    </tr>
{% endfor %}

</table>
<p></p>
<a href="#navigation">top</a>

<!-- 
Check if insert sizes are present, then render the relevant secion
-->
{% if Insertsizes %}
    <div id="insert-sizes"></div>

    <h2>Insert size distribution (Illumina)</h2>
    <div class="infotext">
        Insert size refers to the size of the sequenced DNA fragment. Depending on the exact library protocol, this size will fall fairly uniformly around a mean value (~300-500bp). 
        For Illumina data, that value should typically be (slightly) larger than the combined length of forward and reverse read for optimal data yield. Very flat curves may (depending on the protocol!) 
        indicate a failure during fragment size selection/enrichment. Neither small insert sizes nor flat curves are a clear predictor for subsequent assembly issues, but can inform any potential debugging efforts. 
    </div>
    <p></p>

    {{Insertsizes}}

    <p></p>
    <a href="#navigation">top</a>
{% endif %}

<!--
Busco scores
-->
<div id="busco"></div>
<h2>BUSCO scores</h2>

<div class="infotext">
    <a href="https://busco.ezlab.org/" target="_new">BUSCO</a> scores describe the coverage of the assemblied gene space against a set of broadly conserved singleton genes (here: bacteria_odb10). A perfect assembly should 
    have a complete coverage of the gene space (complete: 100%), without any fragmentation or, worse, duplication. A high value of duplication may indicate assembly errors or contamination. Some taxa with very 
    streamlined gene content, such as <i>Campylobacter</i>, will typically have a completeness score of less than 100%. The Completeness estimates may include duplicated genes, so values greater than 100% are 
    possible (i.e. all genes present, of which x % are duplicated). 
</div>

{{Busco}}

<p></p>
<a href="#navigation">top</a>

<!--
Check if Bracken Illumina data is present, then render the relevant section
-->
{% if Bracken_ILLUMINA %}
    <div id="bracken"></div>
    
    <h2>Bracken - taxonomic composition (ILLUMINA)</h2>

    <div class="infotext">
        <a href="https://github.com/jenniferlu717/Bracken" target="_new">Bracken</a> processes raw outputs from <a href="https://github.com/DerrickWood/kraken2", target="_new">Kraken2</a>, which matches kmers from raw sequencing reads against a reference database to determine the taxonomic composition of a read set. For DNA from
        pure cultures (which is the focus of GABI), only one species should be identified at dominant proportions. For some taxa, like <i>Campylobacter</i>, several species from the same genus may be found at comparative 
        abundances due to a lack of sufficient DNA differences. Otherwise, identification of multiple taxa at higher proportions may indicate a contamination issue.  
    </div>

    {{Bracken_ILLUMINA}}

    <a href="#navigation">top</a>
{% endif %}

<!--
Check if Bracken Nanopore data is present, then render the relevant section 
-->
{% if Bracken_NANOPORE %}
    <div id="bracken_ont"></div>
    <h2>Bracken - taxonomic composition (NANOPORE)</h2>

    <div class="infotext">
        <a href="https://github.com/jenniferlu717/Bracken" target="_new">Bracken</a> processes raw outputs from <a href="https://github.com/DerrickWood/kraken2", target="_new">Kraken2</a>, which matches kmers from raw sequencing reads against a reference database to determine the taxonomic composition of a read set. For DNA from
        pure cultures (which is the focus of GABI), only one species should be identified at dominant proportions. For some taxa, like <i>Campylobacter</i>, several species from the same genus may be found at comparative 
        abundances due to a lack of sufficient DNA differences. Otherwise, identification of multiple taxa at higher proportions may indicate a contamination issue. Please note that Nanopore data is generally of lesser 
        quality than Illumina short reads, so results are to be treated with some caution.
    </div>

    {{ Bracken_NANOPORE}}

    <a href="#navigation">top</a>
{% endif %}

<!--
Check if Bracken Pacbio data is present, then render the relevant section 
-->
{% if Bracken_Pacbio %}
    <div id="bracken_pacbio"></div>
    <h2>Bracken - taxonomic composition (Pacbio)</h2>

    <div class="infotext">
        <a href="https://github.com/jenniferlu717/Bracken" target="_new">Bracken</a> processes raw outputs from <a href="https://github.com/DerrickWood/kraken2", target="_new">Kraken2</a>, which matches kmers from raw sequencing reads against a reference database to determine the taxonomic composition of a read set. For DNA from
        pure cultures (which is the focus of GABI), only one species should be identified at dominant proportions. For some taxa, like <i>Campylobacter</i>, several species from the same genus may be found at comparative 
        abundances due to a lack of sufficient DNA differences. Otherwise, identification of multiple taxa at higher proportions may indicate a contamination issue. 
    </div>

    {{ Bracken_Pacbio}}

    <a href="#navigation">top</a>
{% endif %}

<!--
Sequence characterization
-->

<div id="character"></div>
<h2>Assembly characterization</h2>

<div class="infotest">
    MLST, serotyping and AMR genes
</div>

<p></p>

<table>
<tr>
    <th colspan=1 scope="col" class="grey">Sample</th>
    <th colspan=1 scope="col" class="uneven">Species</th>
    <th colspan=2 scope="col"><div class="tooltip">MLST<span class="tooltiptext">Multi-locus sequence type</span></div></th>
    <th colspan=2 scope="col"><div class="tooltip">AMR<span class="tooltiptext">AMRFinderPlus</span></div></th>
    <th colspan=4 scope="col"><div class="tooltip">Serotype<span class="tooltiptext">Serotype result</span></div></th>
</tr>
<tr>
    <th scope="subcol-inactive"></th>
    <th scope="subcol-inactive"></th>
    <th scope="subcol">Schema</th>
    <th scope="subcol">Type</th>
    <th scope="subcol">Hits</th>
    <th scope="subcol">AMR classes</th>
    <th scope="subcol">Tool</th>
    <th scope="subcol">Serotype</th>
    <th scope="subcol">Patho-Genes</th>
    <th scope="subcol">Pathotype</th>
</tr>
{% for row in summary %}
    <tr class="row-sample">
        <td scope="sample-id">{{row.sample}}</td>
        <td scope="missing">{{row.taxon}}</td>
        <td scope="col_mlst" style="border-left: 2px solid #6e6d6d;">{{row.mlst.scheme}}</td>
        <td scope="col_mlst">{{row.mlst.sequence_type}}</td>
        <td scope="col_amr" style="border-left: 2px solid #6e6d6d;">
            {{row.amrfinder|length}} Entries
        </td>
        <td scope="col_amr">{{row.amrcount}}</td>
        {% if "tool" in row.serotype %}
            <td scope="col_sero" style="border-left: 2px solid #6e6d6d;">{{row.serotype.tool}}</td>
            <td scope="col_sero">{{row.serotype.serotype}}</td>
            <td scope="col_sero">{{','.join(row.serotype.genes)}}</td>
            <td scope="col_sero">{{row.serotype.pathotype}}</td>
        {% else %}
            <td scope="even" style="border-left: 2px solid #6e6d6d;"></td>
            <td scope="even"></td>
            <td scope="even"></td>
            <td scope="even"></td>
        {% endif %}
    </tr>
    <tr class="row-hidden">
        <th colspan=1>Gene</th>
        <th colspan=1><a href=https://github.com/ncbi/amr/wiki/class-subclass target="_new">Class</a></th>
        <th colspan=1><a href=https://github.com/ncbi/amr/wiki/class-subclass target="_new">Subclass</a></th>
        <th colspan=1><a href=https://github.com/ncbi/amr/wiki/Interpreting-results#element-type-and-subtype target="_new">Type</a></th>
        <th colspan=1><a href=https://github.com/ncbi/amr/wiki/Interpreting-results#element-type-and-subtype target="_new">Subtype</a></th>
    </tr>
     {% for entry in row.amrfinder %}
        <tr class="row-hidden">
            <td><div class="tooltip"><a href=https://www.ncbi.nlm.nih.gov/pathogens/isolates/#AMR_genotypes:{{entry.gene_symbol}}* target="_new">{{entry.gene_symbol}}</a><span class="tooltiptext">{{entry.sequence_name}}</span></div></td>
            <td>{{entry.amr_class}}</td>
            <td>{{entry.amr_subclass}}</td>
            <td>{{entry.element_type}}</td>
            <td>{{entry.element_subtype}}</td>
        </tr>
    {% endfor %}
{% endfor %}

</table>

<p></p>
<a href="#navigation">top</a>

<!-- 
Information on used software packages
-->
<div id="software"></div>
<h2>Software versions</h2>

<table class="versions">
    
    {% for pmod,packs in packages.items() %}
        <tr>
            <td scope="col" colspan="2"><b>{{pmod}}</b></td>
        </tr>
        {% for pack in packs %}
            <tr>
                <td></td>
                <td scope="col">{{pack}}</td>
            </tr>
        {% endfor %}
    {% endfor %}
        </table>
<p></p>
<a href="#navigation">top</a>

</div>

<div class="impressum">
    Report generated by bio-raum/gabi. Please check out our <a href="https://github.com/bio-raum/gabi/blob/main/docs/usage.md">documentation</a>. 
</div>
</body>
</html>